### PR TITLE
Fix various bugs with UI rounded borders

### LINKED ISSUE #13503: Ui border is 1 pixel too large, or is getting aliased.
## Bevy version

Main, commit 0ec634763ec378193eeea0cc38909e7d7251b272

## Relevant system information
- Windows 11
- Rust Nightly, 1.80.0 (8387315ab 2024-05-14)
- Also tested on Rust Stable, 1.78.0 (9b00956e5 2024-04-29)
```ignore
AdapterInfo { name: "NVIDIA GeForce RTX 2080 Ti", vendor: 4318, device: 7684, device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "555.85", backend: Vulkan }
```

## What you did

Create a NodeBundle with a 1 pixel wide border on all sides.

## What went wrong

The border appears to be 1 pixel too large on all sides, or is getting aliased, causing it to blend into the node's background. I would expect a 1 pixel border to be exactly 1 pixel on the screen(see the 0.13.2 screenshot below).

[This gist](https://gist.github.com/ImDanTheDev/117be26a909bcc16a49b6247836eb4ed) reproduces the issue on commit 0ec6347.
[And this gist](https://gist.github.com/ImDanTheDev/887e0cc67a66f869e05534f1a7baf1e8) does not show the issue on 0.13.2.

### Possible Fix
I'm not at all fluent with shader code, but while poking around in [ui.wgsl](https://github.com/bevyengine/bevy/blob/0ec634763ec378193eeea0cc38909e7d7251b272/crates/bevy_ui/src/render/ui.wgsl#L161), I found that on line 161, changing `external_distance < internal_distance` to `external_distance > internal_distance` fixes the issue. After doing that, the border appears exactly the same as the 0.13.2 screenshot below. I have no idea what the implications are for making that change, so it may very well break other things.

True scale:
true_scale

Scaled up to highlight issue:
scaled_up

Expected(using bevy 0.13.2):
image

### PR DESCRIPTION
**Pull request by hymm**
# Objective

- Fixes #13503 
- Fix other various bugs I noticed while debugging above issue.

## Solution

- Change the antialiasing(AA) method. It was using fwidth which is the derivative between pixels, but there were a lot of artifacts being added from this. So just use the sdf value. This aa method probably isn't as smooth looking, but better than having artifacts. Below is a visualization of the fwidth.
image
- Use the internal sdf for drawing the background instead of the external sdf and extract the border for these type of nodes. This fixed 2 bugs, one with the background coloring the AA pixels on the edge of rounded borders. And also allows for the border to use a transparent color.
- Don't extract borders if all the widths are zero.

## Testing

- played a bunch with the example in the linked issue.
This PR:
image
Main:
image

- ran the `borders` and `rounded_borders` examples

---

## Changelog

- Fixed various antialiasing issues to do with rounded ui borders.

### CODE REVIEW

File: crates/bevy_ui/src/render/mod.rs [hunk 1]
```diff
@@ -232,6 +237,23 @@ pub fn extract_uinode_background_colors(
             // so we have to divide by `UiScale` to get the size of the UI viewport.
             / ui_scale.0;
 
+        // Both vertical and horizontal percentage border values are calculated based on the width of the parent node
+        // <https://developer.mozilla.org/en-US/docs/Web/CSS/border-width>
```
Comment (Weibye) [inline crates/bevy_ui/src/render/mod.rs:241]: layout and style is a complex topic, having these comments that refers back to source material helps a lot, thanks!

File: crates/bevy_ui/src/render/mod.rs [hunk 2]
```diff
@@ -232,6 +237,23 @@ pub fn extract_uinode_background_colors(
...
+        let left =
+            resolve_border_thickness(style.border.left, parent_width, ui_logical_viewport_size);
+        let right =
+            resolve_border_thickness(style.border.right, parent_width, ui_logical_viewport_size);
+        let top =
+            resolve_border_thickness(style.border.top, parent_width, ui_logical_viewport_size);
+        let bottom =
+            resolve_border_thickness(style.border.bottom, parent_width, ui_logical_viewport_size);
+
+        let border = [left, top, right, bottom];
```
Comment (Weibye) [inline crates/bevy_ui/src/render/mod.rs:255]: All of this makes sense and is straightforward to read 👍

File: crates/bevy_ui/src/render/ui.wgsl [hunk 3]
```diff
@@ -145,32 +151,31 @@ fn draw(in: VertexOutput) -> vec4<f32> {
...
+    return vec4(color.rgb, color.a * t);
+}
 
-    // The item is a rectangle, draw normally with anti-aliasing at the edges.
-    let t = 1. - smoothstep(0.0, fexternal, external_distance);
+fn draw_background(in: VertexOutput) -> vec4<f32> {
+    let texture_color = textureSample(sprite_texture, sprite_sampler, in.uv);
+    let color = select(in.color, in.color * texture_color, enabled(in.flags, TEXTURED));
 
+    /// When drawing the background only draw the internal area and not the border.
```
Comment (Weibye) [inline crates/bevy_ui/src/render/ui.wgsl:168]:
```suggestion
    // When drawing the background only draw the internal area and not the border.
```

**Review by Weibye — APPROVED** [review]
This looks good to me, nice and clean implementation.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**